include/rcutils/macros.h

```diff
@@ -142,6 +142,19 @@ extern "C"
 # define RCUTILS_UNLIKELY(x) (x)
 #endif  // _WIN32
 
+// Provide the compiler a hint about an argument being nonnull when possible.
+#ifndef _WIN32
```

Comment:
So it does look like the NONNULL attributes are supported in gcc and clang, so that's good.  My concern here is with other, non-gcc and clang compilers (maybe what QNX is using, for instance?).  Instead of making this conditional on `!WIN32`, should we make it `GCC || CLANG`?  Not sure.

Reply:
I do not think QNX supports it, i might be wrong. and this is compiler specific.

Reply:
Yes, but I can't know what does and doesn't work on other compilers without testing. I believe this will work on all the compilers we test. If someone is testing with QNX, then they can report it and we can fix it.

My concern with doing something like `GCC || CLANG` is that I will miss some match, and future or older versions of gcc or clang may not match this statement where it should.

This (using `#ifndef _WIN32`) is how we do things in a lot of places already.

It basically boils down to:

- this is how we expect most compilers to work
- except windows... 
- potentially more exceptions in the future

Until I know what the alternative should be (for other compilers) I leave them out for now. 🤷

Reply:
Based on my comment above, and after thinking on it more, I would like to keep it as-is, but please let me know what you guys think.

If you think we really should change it, do you know of an example where we condition on gcc/clang? I was looking in our code base and didn't see it. There are lots of ways suggested online I just don't know which would be best.

Reply:
Looking around a bit, I think that QNX uses gcc under the hood, so it probably understands these attributes.  I also looked around the codebase a bit, and I saw no instances where we were special-casing `__attribute__` for QNX.  The only places I saw special-casing for QNX is around libc-style differences.

But QNX was just an example.  Thinking about it more, there are really only 3 major compilers out there: gcc, clang, and MSVC.  While there are other minor compilers, I doubt this would be the only problem they would have.

All of that is to say; I'm fine with this as-is.

Reply:
Ok, thanks.

I kind of agree, if you want to introduce a new compiler (outside of gcc/clang and msvc) you'll have lots of things to fix. I don't necessarily want to add to that, but on the other hand I don't think I can make it better at the moment.